### ISSUE TITLE
webdav PUT requests are not atomic

### ISSUE THREAD

**Issue by butonic** [cs3org/reva#86]
see eg https://github.com/cs3org/reva/blob/master/cmd/revad/svcs/httpsvcs/ocdavsvc/put.go#L178

first a stat request is made. then the put is made, but between the stat and the put the file could have been overwritten. we need to pass the etag/if match headers to the actual storageprovider or we need to implement locking.

copy has the same problem

cc @DeepDiver1975 @labkode

**Comment by labkode** [comment]
@butonic I agree, this need to be added in the CS3 APIs protocol as first class citizen

**Comment by DeepDiver1975** [comment]
Any news on this? I'd say this is rather critical and requires proper engineering.
What are the options in your opinion @labkode ? THX

**Comment by labkode** [comment]
@DeepDiver1975 this should be possible right now as extra metadata can be added when initiating an upload (mtime, etags) and is up to the storage implementation (fs, eos) to validate these parameters and make the implementation as atomic as possible.

@PVince81 has been working on passing this extra info down to the storage, I've merged his PR today.

**Comment by DeepDiver1975** [comment]
> @DeepDiver1975 this should be possible right now as extra metadata can be added when initiating an upload (mtime, etags) and is up to the storage implementation (fs, eos) to validate these parameters and make the implementation as atomic as possible.

So you say all these header checks *can* be done on storage implementation layer. My expectation would be that this is part of the behavioral spec of the calls. Otherwise we end up in totally different behavior and clients have a hard time to implement proper behavior.

In case you don't want to put all these webdav driven behavior into CS3Api (which I assume ;-) ) we need a way to force atomicity across multiple calls which means an api to lock the files.

**Comment by labkode** [comment]
@DeepDiver1975 how would you do the check on upper layers and make the storage implementation atomic?

**Comment by butonic** [comment]
pointer for the go lockedfile package: https://go.googlesource.com/proposal/+/master/design/33974-add-public-lockedfile-pkg.md

**Comment by refs** [comment]
@butonic we could use https://godoc.org/github.com/rogpeppe/go-internal/lockedfile

**Comment by refs** [comment]
@labkode at first thought, I think a possible mechanism to tackle atomicity on the upper layers would be to introduce transactions to the CS3 API.

Because a webserver could potentially be on a different machine we cannot enforce atomicity on this layer, therefore the implementation details leak one layer above to the API, and we're at the beginning of this discussion where checks are to be made on the storage layer. Passing an ETAG or headers across process boundaries (web-server -> storage) works, but as @DeepDiver1975 said:

> [...] we end up in totally different behavior and clients have a hard time implementing proper behavior.

Which I guess means that storages can choose whether or not to implement atomicity, as of right now this mechanism relies only on interpreting opaque metadata.

Citing @labkode:

> I agree, this need to be added in the CS3 APIs protocol as first class citizen

I see transactions at the API level as a possible way of achieving the goals in this ticket. Of course, this comes with quite a bit of tasks that we could define :)

**Comment by refs** [comment]
Also, @butonic regarding locking we would have to be very careful if we don't have a system like transactions in place and introduce file locking on the storage layer based on ETAG or other metadata. I see the case of a `PUT` request:

### Without transactions:

preconditions: have a gateway client.

1. **`STAT` resource.**
1.1. do defensive code with the result of `STAT`.
2. **upload new resource (`PUT` through the data service).**
3. **`STAT` again** to have the updated metadata.

Any locking going on here, since it is on a different layer (WebDAV) might result on the storage locking a resource but if a request fails along the way, that resource is never freed potentially until a timeout. With scripted transactions, we would be able to go around that problem:

### With transactions:

preconditions: have a gateway client.

1. prepare a transaction on the WebDAV service.
2. encapsulate steps 1, 2 and 3 from the previous use case on the transaction.
3. execute it through the gateway client.

This way the storage layer has an entire plan of the series of steps that have to be executed, if something fails along the way is not up to the transport layer but all is within the context of the storage. This also roughly translates into ease to free resources and less locking between goroutines.

There is still the case of the first `STAT` the WebDAV layer does in order to perform some defensive code (i.e validating etags):

https://github.com/cs3org/reva/blob/6844ee0648712c9b2f9d868b977da6b5b3bb0f1d/internal/http/services/owncloud/ocdav/put.go#L181-L189

This can also be delegated to the storage as part of the transaction. In fact, this check should be a pre-requisite in order to perform a file update, and the fact that we have it on the WebDAV layer is sub-optimal.

I guess all I'm proposing here is that operations that require modifying files have to be wrapped in a transaction.

/cc @labkode @butonic @DeepDiver1975 @kulmann @micbar for opinions.

**Comment by butonic** [comment]
Unfortunately, HTTP is a stateless protocol. Ultimately, the PATCH/POST/PUT request to the dataprovider needs to be atomic. Which requires sending an `If-Match` header and checking it when processing the upload. AFAICT we need to check `If-Match` in `FinalizeUpload()`. If the etag still maches, we can start moving the file to final destination.

AFAIR we currently always stat the target file because need to return different HTTP statuses dependting on wether the file existed or not. Maybe we can omit that if the request contains the `If-Match` header. It would be great if the web ui was sending the `If-Match` header as well.

**Comment by refs** [comment]
Yes, I'm proposing shifting responsibilities of atomicity down to the storage, since as you said HTTP is stateless. Having a plan for the transaction, the storage would then take action and do the steps as planned. This shifts the responsibility of having to STAT -> upload -> STAT from WebDAV to the storage. The WebDAV can still prepare a transaction for the storage, even request the upload endpoint and add it to a (hypothetically) `Transaction` struct (that is part of the CS3 API) and the storage would need to be extended so it can get transactions. So the entire state is encapsulated in this Transaction object, that is included on the cs3api. Just to be clear, a transaction is naturally defined as a protobuf spec.
